api/src/main/java/org/openmrs/module/cohort/CohortM.java

```diff
@@ -423,4 +432,72 @@ public void setId(Integer cohortId) {
 	public String toString() {
 		return this.name;
 	}
+	
+	public String getDefinitionHandlerClassname() {
+		return definitionHandlerClassname;
+	}
+	
+	public void setDefinitionHandlerClassname(String definitionHandlerClassname) {
+		this.definitionHandlerClassname = definitionHandlerClassname;
+	}
+	
+	public String getDefinitionHandlerConfig() {
+		return definitionHandlerConfig;
+	}
+	
+	public void setDefinitionHandlerConfig(String definitionHandlerConfig) {
+		this.definitionHandlerConfig = definitionHandlerConfig;
+	}
+	
+	public void addMembership(CohortMember cohortMember) {
+		CohortDefinitionHandler cohortDefinition = getDefinitionHandler();
+		if (cohortDefinition instanceof ManualCohortDefinitionHandler) {
+			((ManualCohortDefinitionHandler) cohortDefinition).addMember(this, cohortMember);
+		} else {
+			throw new ManualChangeNotSupportedException("Manual changes to the cohort aren't supported");
+		}
+	}
+
+	public void removeMembership(CohortMember cohortMember) {
+		if (getDefinitionHandlerClassname().isEmpty()) {
+			throw new APIException("CohortDefinitionHandler is null");
+		}
+		CohortDefinitionHandler cohortDefinition = getDefinitionHandler();
+		if (cohortDefinition instanceof ManualCohortDefinitionHandler) {
+			((ManualCohortDefinitionHandler) cohortDefinition).removeMember(this, cohortMember);
+		} else {
+			throw new ManualChangeNotSupportedException("Manual changes to the cohort aren't supported");
+		}
+	}
+
+	/**
+	 * Loads cohort definition handler class - If the cohortDefinitionHandlerClassname is null, use
+	 * defaultCohortDefinitionHandler
+	 *
+	 * @return CohortDefinitionHandler
+	 */
+	public CohortDefinitionHandler getDefinitionHandler() {
```

Comment:
@ibacher 
Is this the right place to be doing?
What's the right approach for loading handler classes?

Reply:
That's a good question...

I like the idea of checking with Spring. Maybe it's better to do:

```java
Class<?> definitionHandlerClass;
try {
	definitionHandlerClass = Context.loadClass(getDefinitionHandlerClassname());
} catch (InstantiationException | ClassNotFoundException | IllegalAccessException e) {
	// throw exception
}

if (!(cohortDefinitionHandler instanceof CohortDefinitionHandler)) {
	// throw exception
}

List<? extends CohortDefinitionHandler> handlers = Context.getRegisteredComponents(definitionHandlerClass);

if (!handlers.isEmpty()) {
	return handler.get(0);
}

return definitionHandlerClass.getDeclaredConstructor().newInstance();
```

We probably also need a validator here to ensure that a Cohort always has a `definitionHandlerClassname` property.

Re: Where this goes, I think in general, we've tended to locate code like this as a utility method to keep the implementation of the domain object fairly simple. Then again, it's a little absurd to have a utility class that is then used by this class, so I think it's fine to be here for now.

Reply:
Make sense to me